66 timetable model converter

Firebrandt opened the pull request:
Made a TimetableModelUnpacker class that facilitates unpacking our Model classes into their original entities. Also made a suitable test class. This required some smaller changes throughout, such as implementing equals() methods (to compare unpacked and expected Entities) for Session, Course, etc. and adding an extra field to TimetableModel to better model the Timetable class.

Overall Project Impact:
- Should provide somewhat convenient support for teammates when writing their code; a teammate mentioned in the Discord that something to make converting between Models and Entities easier would be useful, e.g. for repacking into JSON. @moonsdust 👀

- Small edits made to other classes that were done to support the writing of unit tests for the unpacker, should make writing future unit tests that involve comparing entities easier.
 
Design decisions:
- Implemented Singleton Design Pattern since I believe we would never want to initialize more than one instance of this class. Although it's pretty light weight, this is probably still good for screaming architecture purposes.
- This is very minor but I factored out a small method when I realized it was getting too long and it could end up being repeat code, so I avoided that code smell?
- Single Responsibility Principle: You can probably see from the title of this issue that it was really tempting to try and put this in the TimetableModelConverter class, which already exists. However, 2 different types of conversion means the class would be reponsible to people that want to convert to a [X_X]ViewModel object (and any changes they'd want to make to the class), and whoever wants the original Entity object (and any changes they'd want). So I made a new one, I hope that's justified.

Please provide feedback on:
- The FIXME I put in - there are 2 nearly identical methods with one line changed. I think it's a code smell but I'm not quite sure how I'd go about fixing it. Will check on RefactorGuru but LMK if you come up with anything.
- Any other code smells you see :)

Firebrandt committed:
Wrote a draft of the TimetableModelUnpacker class.

Firebrandt committed:
Drafted some unit tests for the unpacker class. Will need to debug.

Firebrandt committed:
Added equals() method for Course, overrides the obj one.
Added equals() for Session and Timetable too, to support the unit tests (which otherwise break).
Equality works the same as a Python dataclass.

Added a SessionType field to TimetableModel to support conversion between the two.

michaelhsj commented on src/main/java/overlap_use_case/TestTimetableModelUnpacker.java:
```diff
@@ -0,0 +1,101 @@
+
+    /** Test unpackBlockModel with a basic case. **/
+    @Test
+    public void test_unpack_block_model_basic(){
```
I would rename these methods so that they fall in line with our camelCaseMethodName conventions. Overall test coverage here looks good!

Firebrandt replied:
Done!

michaelhsj commented on src/main/java/overlap_use_case/TimetableModelUnpacker.java:
```diff
@@ -0,0 +1,95 @@
+        return session;
+    }
+
+    public static Timetable unpackTimeTable(TimetableModel timetable){
```
Just a tiny little correction so that it falls in line with the name of the Timetable class: I would call it unpackTimetable.

Firebrandt replied:
AUUGH (done!)

michaelhsj commented on src/main/java/overlap_use_case/TimetableModelUnpacker.java:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    // FIXME: I think this might be a code smell. Two nearly identical methods with one slightly changed constructor call.
```
If you have a bunch of repeated code between unpackCourseModelAsCalendarCourse and unpackCourseModelasTimetableCourse, it may help to write a private helper method that completes the shared code (if the shared code is particularly long). Otherwise, this seems like the best we can do right now.

Firebrandt replied:
Well, I pulled out the loop. But it's literally just unpackSection() over a list, and while it doesn't hurt to factor it out, it is just one loop.

The rest of the method is pretty much just a return statement, which can't be usefully simplified imo. 

Thanx tho! :3

michaelhsj commented on src/main/java/overlap_use_case/TimetableModelUnpacker.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    public static Block unpackBlockModel(BlockModel block){
+        String blockDay = switch (block.getDay()) {
```
It looks like this particular set of switch case statements is breaking the autograder. Is there any way for you to refactor it into something the autograder will accept without raising an IntelliJ warning?

Firebrandt replied:
Tried it! Strange since this is just syntax sugar (it seems like a valid expression, at least in the version of Java we're using!).
Hopefully this more conventional version holds :D.

michaelhsj commented on src/main/java/overlap_use_case/TimetableModelUnpacker.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class TimetableModelUnpacker {
+
+    public static Session unpackSessionModel(SessionModel sessionModel){
```
I think it would be appropriate to write some javadoc for the methods in the TimetableModelUnpacker, since every non-trivial method should have some sort of documentation.

Firebrandt replied:
Tried it! Strange since this is just syntax sugar (it seems like a valid expression, at least in the version of Java we're using!). 
Hopefully this more conventional version holds :D.

Firebrandt replied:
Also done! :)

Firebrandt committed:
Implemented Michael H's revisions. Added additional documentation and small formatting edits to code.

Also improved the function of reformatModelTime since we decided on a change in the Discord.

Also handled the IntelliJ warnings.

Firebrandt committed:
Made the switch case into an if because the autograder sucks.

Firebrandt commented:
Something is really off with the Auto-grader. My IDE compiles! D: Help?

YahyaElgabra committed:
implemented TimetablesSortInteractor, documented my code

next steps: integrate with mainmenu and timetableUI, flow, tests

YahyaElgabra committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-13-finishing-sorter

YahyaElgabra committed:
checking if github still hates enhanced for loops

YahyaElgabra committed:
Implemented flow publisher, still didn't integrate it with the rest of my use case

YahyaElgabra committed:
Merge branch 'main' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
no implementations yet, stuck on how to convert TimetableViewModel into TimetableModel

YahyaElgabra committed:
Merge branch 'feature-9-finish-main-menu-ui' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
I will go to WAR with swing

WHY CANT IT JUST LET ME SWITCH SCREENS AAAAAAAAAAAAAAH

YahyaElgabra committed:
started doing tests but stopped cuz i need sleep
for mainmenu integration: ask michael about swing
for ttui integration: ask people about converting from timetableviewmodel down
for flow: done

Firebrandt committed:
DESIGN DECISION MADE:
- UNDID THE OBSERVER PATTERN SINCE I WAS RUNNING INTO A KIND OF OVERCOMPLICATING IT CODE SMELL. REFACTORGURU MENTIONS IT, BUT BASICALLY, THE PATTERN OVERCOMPLICATES THINGS AND ACTUALLY MAKES IT HARDER.

- More refactoring since I found out some clean architecture was wrong. Turns out I need to pass in ViewModel to the InputDialog, so I've made those too.

- Added some skeletal support for TimetableUI from Kai. Hopefully this is good!

Firebrandt committed:
Factored out another CA violation into yet another convert! !!!
!!! !!! :D

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
observer pattern stuff

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java

Firebrandt committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-3-overlapping-shenanigans

# Conflicts:
#	build.gradle
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableUI.java
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableViewBlockModel.java
#	src/main/java/screens/MainUI.java

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Made a kind of mechanism to wait for JD's thing to finish. Ready to try running it now... I guess?

Firebrandt committed:
Set up a draft for my initialization code.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Commit just to save

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

Firebrandt committed:
Fixed architecture problem with the Controller (output goes through a Presenter to the View now :) )

Firebrandt committed:
Fixed unit tests for one of the Converters.

Firebrandt committed:
Made unit tests for the other converter.

Firebrandt committed:
Fixed my interactor unit tests >:D

Firebrandt committed:
Made new unit tests. Doing super integration now.

Refactored to not kill the main frame (instead refactors it).

Firebrandt committed:
Fixed my file layout.

Firebrandt committed:
Added a new function to update UI to the InputDialog.

Firebrandt committed:
Refined the UI resetting stuff a little bit.

Firebrandt committed:
Implement Michael H's revisions. Fixed warnings in my files.

Added a strings.properties file to store UI strings. Centralizing them in one place could be a good idea to increase UI maintainability (what if we need to translate this abominable application into a different language?)

Firebrandt committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java
#	src/main/java/display_timetable_use_case/frameworks_and_drivers/TimetableUI.java
#	src/main/java/display_timetable_use_case/frameworks_and_drivers/TimetableViewBlockModel.java
#	src/main/java/display_timetable_use_case/interface_adapters/DisplayTimetableController.java
#	src/main/java/edit_timetable_use_case/frameworks_and_drivers/AddSectionsMenu.java
#	src/main/java/fileio_use_case/frameworks_and_drivers/TimetableGateway.java
#	src/main/java/overlap_crap_fix_locations_later/OverlapInputDialog.java
#	src/main/java/screens/MainUI.java
#	src/main/java/timetables_sort_use_case/frameworks_and_drivers/AllTimetablesScreen.java
#	src/test/java/overlap_stuff_fix_layout_later/OverlapInputDialogTest.java

Firebrandt committed:
Refactored the equals methods to be a little less terrible and hopefully not break autograder.

Firebrandt committed:
Merge branch 'main' into 66-TimetableModelConverter

# Conflicts:
#	src/main/java/entities/Block.java
#	src/main/java/entities/Section.java
#	src/main/java/entities/Timetable.java
#	src/main/java/entities/TimetableCourse.java
#	src/main/java/retrieve_timetable_use_case/RetrieveTimetableInteractor.java
#	src/main/java/retrieve_timetable_use_case/TimetableModel.java
#	src/main/java/retrieve_timetable_use_case/application_business/EntityConverter.java
#	src/main/java/retrieve_timetable_use_case/interface_adapters/TimetableModelConverter.java

michaelhsj reviewed (commented):
Looks good! Nothing's really changed since the last commit except for clearing out some IntelliJ warnings, so you're good to go.

michaelhsj reviewed (approved):
Looks good! Nothing's really changed since the last commit except for clearing out some IntelliJ warnings, so you're good to go.
